go/src/github.com/mesosphere/rendler/scheduler/main.go

```diff
@@ -1,281 +1,284 @@
+	task := s.makeTaskPrototype(offer)
+	task.Name = proto.String("CRAWL_" + *task.TaskId.Value)
+	task.Executor = s.crawlExecutor
+	task.Data = []byte(url)
+	return task
+}
 
-	maxTasksForOffer := func(offer mesos.Offer) int {
-		// TODO(nnielsen): Parse offer resources.
-		count := 0
+func (s *RendlerScheduler) makeRenderTask(url string, offer *mesos.Offer) *mesos.TaskInfo {
+	task := s.makeTaskPrototype(offer)
+	task.Name = proto.String("RENDER_" + *task.TaskId.Value)
+	task.Executor = s.renderExecutor
+	task.Data = []byte(url)
+	return task
+}
 
-		var cpus float64 = 0
-		var mem float64 = 0
+func (s *RendlerScheduler) Registered(
+	driver sched.SchedulerDriver,
+	frameworkId *mesos.FrameworkID,
+	masterInfo *mesos.MasterInfo) {
+	log.Printf("Registered")
+}
 
-		for _, resource := range offer.Resources {
-			if resource.GetName() == "cpus" {
-				cpus = *resource.GetScalar().Value
-			}
+func (s *RendlerScheduler) Reregistered(driver sched.SchedulerDriver, masterInfo *mesos.MasterInfo) {
+	log.Println("Framework Re-Registered with Master ", masterInfo)
+}
 
-			if resource.GetName() == "mem" {
-				mem = *resource.GetScalar().Value
+func (s *RendlerScheduler) Disconnected(sched.SchedulerDriver) {
+	log.Println("Framework disconnected with Master")
+}
+
+func (s *RendlerScheduler) ResourceOffers(driver sched.SchedulerDriver, offers []*mesos.Offer) {
+	log.Printf("Received %d resource offers", len(offers))
+	for _, offer := range offers {
+		select {
+		case <-s.shutdown:
+			log.Println("Shutting down: declining offer on [", offer.Hostname, "]")
+			driver.DeclineOffer(offer.Id, DEFAULT_FILTER)
+			continue
+		default:
+			tasks := []*mesos.TaskInfo{}
+
+			for i := 0; i < maxTasksForOffer(offer)/2; i++ {
+				if s.crawlQueue.Front() != nil {
+					url := s.crawlQueue.Front().Value.(string)
+					s.crawlQueue.Remove(s.crawlQueue.Front())
+					task := s.makeCrawlTask(url, offer)
+					tasks = append(tasks, task)
+				}
+				if s.renderQueue.Front() != nil {
+					url := s.renderQueue.Front().Value.(string)
+					s.renderQueue.Remove(s.renderQueue.Front())
+					task := s.makeRenderTask(url, offer)
+					tasks = append(tasks, task)
+				}
 			}
-		}
 
-		for cpus >= TASK_CPUS && mem >= TASK_MEM {
-			count++
-			cpus -= TASK_CPUS
-			mem -= TASK_MEM
+			if len(tasks) == 0 {
+				driver.DeclineOffer(offer.Id, DEFAULT_FILTER)
+			} else {
+				driver.LaunchTasks([]*mesos.OfferID{offer.Id}, tasks, DEFAULT_FILTER)
+			}
 		}
-
-		return count
 	}
+}
 
-	printQueueStatistics := func() {
-		// TODO(nnielsen): Print queue lengths.
+func (s *RendlerScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos.TaskStatus) {
+	log.Printf("Received task status [%s] for task [%s]", rendler.NameFor(status.State), *status.TaskId.Value)
+
+	if *status.State == mesos.TaskState_TASK_RUNNING {
+		s.tasksRunning++
+	} else if rendler.IsTerminal(status.State) {
+		s.tasksRunning--
+		if s.tasksRunning == 0 {
```

Comment:
Maybe we won't gain so much in this particular example but it's good to reduce nesting as much as possible as it reduces the number of scope levels you have to hold in your head.
```go
func (s *RendlerScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos.TaskStatus) {
	log.Printf("Received task status [%s] for task [%s]", rendler.NameFor(status.State), *status.TaskId.Value)

	if *status.State == mesos.TaskState_TASK_RUNNING {
		s.tasksRunning++
	} else if rendler.IsTerminal(status.State) {
		s.tasksRunning--
	}

    if s.tasksRunning > 0 {
        return
    }

    select {
    case <-s.shutdown:
        close(s.done)
    default:
    }
}
```